routing/payment_lifecycle.go

```diff
@@ -319,27 +322,42 @@ lifecycle:
 }
 
 // checkTimeout checks whether the payment has reached its timeout.
-func (p *paymentLifecycle) checkTimeout() error {
-	select {
-	case <-p.timeoutChan:
-		log.Warnf("payment attempt not completed before timeout")
-
+func (p *paymentLifecycle) checkTimeoutAndContext(ctx context.Context) error {
+	failPayment := func(reason channeldb.FailureReason) error {
 		// By marking the payment failed, depending on whether it has
 		// inflight HTLCs or not, its status will now either be
 		// `StatusInflight` or `StatusFailed`. In either case, no more
 		// HTLCs will be attempted.
-		err := p.router.cfg.Control.FailPayment(
-			p.identifier, channeldb.FailureReasonTimeout,
-		)
+		err := p.router.cfg.Control.FailPayment(p.identifier, reason)
 		if err != nil {
 			return fmt.Errorf("FailPayment got %w", err)
 		}
 
+		return nil
+	}
+
+	select {
+	case <-p.timeoutChan:
+		log.Warnf("payment attempt not completed before timeout")
+		err := failPayment(channeldb.FailureReasonTimeout)
+		if err != nil {
+			return err
+		}
+
+	case <-ctx.Done():
+		log.Warnf("payment attempt context canceled")
+		err := failPayment(channeldb.FailureReasonError)
+		if err != nil {
+			return err
+		}
+
+		return ctx.Err()
```

Comment:
Unfortunately we cannot return an error here, as it'd cause the payment loop to exit. For the timeout case, we never exit the loop. Consider this case, say a payment is made of two HTLC attempts,
```
HTLC1 ---> hop1 ---> remote
HTLC2 ---> hop2 ---> remote
```
Then remote settles the invoice, but hop2 is offline, we'd end up in this case,
```
HTLC1 ---> hop1 ---> remote ---> settled
HTLC2 ---> hop2 ---> remote ---> pending
```
After 60s, we'd time out this payment, but we won't quit the loop as we still think the payment is inflight,
https://github.com/lightningnetwork/lnd/blob/9d358bc64934caef724108fc9704097db31658c4/channeldb/payment_status.go#L148

Then hop2 comes online and settles it, we'd consider the payment successful.
https://github.com/lightningnetwork/lnd/blob/9d358bc64934caef724108fc9704097db31658c4/channeldb/payment_status.go#L156

Reply:
In other words, once the HTLC is sent, there's no easy way to cancel it unless you get a clear response from your peer saying it's failed/settled. If we exit here because the context is canceled, the payment would still have the status `StatusInFlight`.

Reply:
Thank you for the detailed explanation. I see why we can't interrupt the entire payment loop but instead behave exactly how we do it the timeout case.